File: backend/src/models/Wallet.js
const mongoose = require('mongoose');

const WalletSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    unique: true,
    index: true
  },
  
  balance: {
    type: Number,
    default: 0,
    min: 0,
    get: v => parseFloat(v.toFixed(2))
  },
  
  currency: {
    type: String,
    default: 'USD',
    enum: ['USD', 'EUR', 'GBP', 'coins']
  },
  
  // Earnings breakdown
  totalEarnings: {
    type: Number,
    default: 0,
    min: 0
  },
  pendingEarnings: {
    type: Number,
    default: 0,
    min: 0
  },
  withdrawableBalance: {
    type: Number,
    default: 0,
    min: 0
  },
  lifetimeWithdrawals: {
    type: Number,
    default: 0
  },
  
  // Payment methods
  paymentMethods: [{
    type: {
      type: String,
      enum: ['card', 'paypal', 'bank', 'stripe']
    },
    details: mongoose.Schema.Types.Mixed,
    isDefault: Boolean,
    isVerified: Boolean,
    addedAt: Date
  }],
  
  // Wallet status
  isLocked: {
    type: Boolean,
    default: false
  },
  lockReason: String,
  
  // Last transaction
  lastTransactionAt: Date,
  lastWithdrawalAt: Date,
  
}, {
  timestamps: true,
  toJSON: { getters: true },
  toObject: { getters: true }
});

// Index for lookups
WalletSchema.index({ userId: 1 }, { unique: true });

// Method to add funds
WalletSchema.methods.addFunds = async function(amount, description = '') {
  if (amount <= 0) {
    throw new Error('Amount must be positive');
  }
  
  this.balance += amount;
  this.lastTransactionAt = new Date();
  
  await this.save();
  
  // Create transaction record
  const Transaction = mongoose.model('Transaction');
  await Transaction.create({
    userId: this.userId,
    walletId: this._id,
    type: 'purchase',
    amount,
    description,
    balanceBefore: this.balance - amount,
    balanceAfter: this.balance,
    status: 'completed'
  });
  
  return this;
};

// Method to deduct funds
WalletSchema.methods.deductFunds = async function(amount, description = '') {
  if (amount <= 0) {
    throw new Error('Amount must be positive');
  }
  
  if (this.balance < amount) {
    throw new Error('Insufficient balance');
  }
  
  if (this.isLocked) {
    throw new Error('Wallet is locked');
  }
  
  this.balance -= amount;
  this.lastTransactionAt = new Date();
  
  await this.save();
  
  // Create transaction record
  const Transaction = mongoose.model('Transaction');
  await Transaction.create({
    userId: this.userId,
    walletId: this._id,
    type: 'purchase',
    amount: -amount,
    description,
    balanceBefore: this.balance + amount,
    balanceAfter: this.balance,
    status: 'completed'
  });
  
  return this;
};

// Method to transfer funds to another wallet
WalletSchema.methods.transferTo = async function(recipientWalletId, amount, description = '') {
  const recipientWallet = await mongoose.model('Wallet').findById(recipientWalletId);
  
  if (!recipientWallet) {
    throw new Error('Recipient wallet not found');
  }
  
  await this.deductFunds(amount, `Transfer to ${recipientWallet.userId}: ${description}`);
  await recipientWallet.addFunds(amount, `Transfer from ${this.userId}: ${description}`);
  
  return { sender: this, recipient: recipientWallet };
};

const Wallet = mongoose.model('Wallet', WalletSchema);

module.exports = Wallet;

File: flutter_app/lib/features/wallet/data/models/wallet.dart
class Wallet {
  final String id;
  final String userId;
  final double balance;
  final int coins;
  final String currency;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Wallet({
    required this.id,
    required this.userId,
    this.balance = 0.0,
    this.coins = 0,
    this.currency = 'USD',
    this.createdAt,
    this.updatedAt,
  });

  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      id: json['_id'] ?? json['id'] ?? '',
      userId: json['userId'] ?? json['user'] ?? '',
      balance: (json['balance'] ?? 0).toDouble(),
      coins: json['coins'] ?? 0,
      currency: json['currency'] ?? 'USD',
      createdAt:
          json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt:
          json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'balance': balance,
      'coins': coins,
      'currency': currency,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }
}

class Transaction {
  final String id;
  final String walletId;
  final String type; // credit, debit, transfer, gift, purchase, topup
  final double amount;
  final int? coins;
  final String status; // pending, completed, failed, cancelled
  final String? description;
  final String? fromUserId;
  final String? toUserId;
  final String? referenceId;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;

  Transaction({
    required this.id,
    required this.walletId,
    required this.type,
    required this.amount,
    this.coins,
    this.status = 'completed',
    this.description,
    this.fromUserId,
    this.toUserId,
    this.referenceId,
    this.metadata,
    this.createdAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['_id'] ?? json['id'] ?? '',
      walletId: json['walletId'] ?? json['wallet'] ?? '',
      type: json['type'] ?? 'credit',
      amount: (json['amount'] ?? 0).toDouble(),
      coins: json['coins'],
      status: json['status'] ?? 'completed',
      description: json['description'],
      fromUserId: json['fromUserId'] ?? json['from'],
      toUserId: json['toUserId'] ?? json['to'],
      referenceId: json['referenceId'],
      metadata: json['metadata'],
      createdAt:
          json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'walletId': walletId,
      'type': type,
      'amount': amount,
      if (coins != null) 'coins': coins,
      'status': status,
      if (description != null) 'description': description,
      if (fromUserId != null) 'fromUserId': fromUserId,
      if (toUserId != null) 'toUserId': toUserId,
      if (referenceId != null) 'referenceId': referenceId,
      if (metadata != null) 'metadata': metadata,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
    };
  }

  bool get isCredit => type == 'credit' || type == 'topup' || type == 'gift';
  bool get isDebit =>
      type == 'debit' || type == 'transfer' || type == 'purchase';
}
